apps/docs/refactor.md

```diff
@@ -0,0 +1,565 @@
+the global redux state tree. https://github.com/erikras/ducks-modular-redux
+describes one pattern for doing this, which we will riff on.
+
+Feature libraries that expose shared redux state should have the following
+directory structure:
+
+```
+src/
+  <feature-name>/
+    index.js
+    redux/
+      reducer.js
+      actionTypes.js
+      actionCreators.js
+      selectors.js
+```
+
+In this hypothetical feature, we expose the state and actions for a stepper
+feature for use in other libraries. In order to keep encapsulation high, we want
+to only expose the action types, action creators, and selectors that other
+libraries need to use, while keeping the rest more "private" to our own
+library. Also, it's important to consider that other code which uses this feature
+library should not have to know about the shape of the state provided by the
+feature. Likewise, the feature should not have to know how it's state gets
+stored in the global redux state tree.
+
+Here is how the above files might look if we wanted to expose only step
+incrementation, but not step decrementation.
+
+**`src/stepper/redux/actionTypes.js`**
+
+Action types are prefixed with the name of the feature to avoid collisions with
+other libraries and their reducers.
+
+```js
+export const INCREMENT_STEP = "stepper/INCREMENT_STEP";
+export const DECREMENT_STEP = "stepper/DECREMENT_STEP";
+```
+
+**`src/stepper/redux/reducer.js`**
+
+The reducer for the feature library should operate on a subtree of the global
+state tree, rather than on the entire state tree. Note that this library acts
+like a singleton in the sense that there is only one corresponding state object
+for the library on the page at any given time.
+
+```js
+import * as actionTypes from './actionTypes';
+
+export default reducer(state={step: 1}, action) {
+  switch (action.type) {
+    case actionTypes.INCREMENT_STEP:
+      return {step: state.step + 1};
+    case actionTypes.DECREMENT_STEP:
+      return {step: state.step - 1};
+    default:
+      return state;
+  }
+}
+```
+
+**`src/stepper/redux/actionCreators.js`**
+
+Action creators should make sure to generate actions with types that are
+globally unique on the page.
+
+```js
+import * as actionTypes from './actionTypes';
+
+export function incrementStep() {
+  return {type: actionTypes.INCREMENT_STEP}
+}
+
+export function decrementStep() {
+  return {type: actionTypes.DECREMENT_STEP}
+}
+```
+
+**`src/stepper/redux/selectors.js`**
+
+Selector functions for the feature must be "configured" before they can be used
+so that they know how to access the part of the global state tree with the data
+they are interested in. This is done using the `setRootSelector` function.
+
+```js
+let getRoot = () => throw new Error("rootSelector not set. Did you forget to call configure?");
+
+export function setRootSelector(rootSelector) {
+  getRoot = rootSelector;
+}
+
+export function getStep(state) {
+  return getRoot(state).step;
```

Comment:
This approach means the stepper feature still needs to now how it fits into the larger library. Wouldn't it be better to just have the caller tell it this?

i.e.
```
// selectors.js
let cachedSelector = () => throw new Error('nope');

export function setSelector(selector) {
  cachedSelector = selector;
}

export function getStep(state) {
  return cachedSelector(state);
}

// caller.js
stepper.setSelector(state => state.step);
```